R/two.step.hybrid.R

```diff
@@ -1,170 +1,412 @@
+    cols_to_pivot_rt <- colnames(final.times)[-c(1:4)]
+
+    final.ftrs <- tidyr::pivot_longer(final.ftrs,
+      cols = all_of(cols_to_pivot_int),
+      names_to = "sample",
+      values_to = "sample_intensity"
+    )
+
+    final.times <- tidyr::pivot_longer(final.times,
+      cols = all_of(cols_to_pivot_rt),
+      names_to = "sample",
+      values_to = "sample_rt"
+    )
+
+    recovered_feature_sample_table <- dplyr::full_join(final.times, final.ftrs,
+      by = c("mz", "time", "mz_min", "mz_max", "sample")
+    )
+
+    colnames(recovered_feature_sample_table)[2] <- "rt"
+
+    batchwise[[batch]]$recovered_feature_sample_table <- recovered_feature_sample_table
+
+    batchwise[[batch]]$corrected_features <- batchwise[[batch]]$features2
+    batchwise[[batch]]$extracted_features <- batchwise[[batch]]$features
+  }
+
+  return(batchwise)
+}
+
+two.step.hybrid <- function(filenames,
+                            metadata,
+                            work_dir,
+                            min.within.batch.prop.detect = 0.1,
+                            min.within.batch.prop.report = 0.5,
+                            min.batch.prop = 0.5,
+                            batch.align.mz.tol = 1e-5,
+                            batch.align.chr.tol = 50,
+                            known.table = NA,
+                            cluster = 4,
+                            min.pres = 0.5,
+                            min.run = 12,
+                            mz.tol = 1e-5,
+                            baseline.correct.noise.percentile = 0.05,
+                            shape.model = "bi-Gaussian",
+                            baseline.correct = 0,
+                            peak.estim.method = "moment",
+                            min.bw = NA,
+                            max.bw = NA,
+                            sd.cut = c(0.1, 100),
+                            sigma.ratio.lim = c(0.05, 20),
+                            component.eliminate = 0.01,
+                            moment.power = 2,
+                            align.mz.tol = NA,
+                            align.chr.tol = NA,
+                            max.align.mz.diff = 0.01,
+                            pre.process = FALSE,
+                            recover.mz.range = NA,
+                            recover.chr.range = NA,
+                            use.observed.range = TRUE,
+                            match.tol.ppm = NA,
+                            new.feature.min.count = 2,
+                            recover.min.count = 3,
+                            intensity.weighted = FALSE,
+                            BIC.factor = 2) {
+  metadata <- read.table(metadata, sep = ",", header = TRUE)
+  filenames_batchwise <- bind_batch_label_column(filenames, metadata)
+  batches_idx <- unique(metadata$batch)
+  batchwise <- new("list")
+  message("* processing ", length(batches_idx), " batches separately")
+
+  for (batch.i in 1:length(batches_idx)) {
```

Comment:
There are functions to groupwise compute something in a dataframe, see [here](https://github.com/RECETOX/recetox-xMSannotator/blob/b99fa2873b9540fc297e3fa61b784ebf110d4b7f/xmsannotator/R/compute_rt_modules.R#L112).

Could allow a more elegant solution. Also, if someone now starts counting batches at 0, we would get an error, or if actually the data contained only batches 1 and 3

Reply:
You could also use a normal for loop `for (batch_id in batches_idx)`

Reply:
Nice catch! I actually use `for (batch_id in batches_idx)` in all other loops. Missed this one, fixed in 000aa86.